mailbox/tools/quota-recompute/src/main/java/org/apache/james/mailbox/quota/task/RecomputeCurrentQuotasService.java

```diff
@@ -149,36 +165,49 @@ public Snapshot snapshot() {
     private final UserQuotaRootResolver userQuotaRootResolver;
     private final SessionProvider sessionProvider;
     private final MailboxManager mailboxManager;
+    private final JMAPCurrentUploadUsageRecomputator jmapCurrentUploadUsageRecomputator;
 
     @Inject
     public RecomputeCurrentQuotasService(UsersRepository usersRepository,
                                          CurrentQuotaManager storeCurrentQuotaManager,
                                          CurrentQuotaCalculator currentQuotaCalculator,
                                          UserQuotaRootResolver userQuotaRootResolver,
                                          SessionProvider sessionProvider,
-                                         MailboxManager mailboxManager) {
+                                         MailboxManager mailboxManager,
+                                         JMAPCurrentUploadUsageRecomputator jmapCurrentUploadUsageRecomputator) {
         this.usersRepository = usersRepository;
         this.storeCurrentQuotaManager = storeCurrentQuotaManager;
         this.currentQuotaCalculator = currentQuotaCalculator;
         this.userQuotaRootResolver = userQuotaRootResolver;
         this.sessionProvider = sessionProvider;
         this.mailboxManager = mailboxManager;
+        this.jmapCurrentUploadUsageRecomputator = jmapCurrentUploadUsageRecomputator;
     }
 
     public Mono<Task.Result> recomputeCurrentQuotas(Context context, RunningOptions runningOptions) {
         return Flux.from(usersRepository.listReactive())
             .transform(ReactorUtils.<Username, Task.Result>throttle()
                 .elements(runningOptions.getUsersPerSecond())
                 .per(Duration.ofSeconds(1))
-                .forOperation(username -> recomputeUserCurrentQuotas(context, username)))
+                .forOperation(username -> recomputeQuotasOfUser(runningOptions.getQuotaComponents(), context, username)))
             .reduce(Task.Result.COMPLETED, Task::combine)
             .onErrorResume(UsersRepositoryException.class, e -> {
                 LOGGER.error("Error while accessing users from repository", e);
                 return Mono.just(Task.Result.PARTIAL);
             });
     }
 
-    private Mono<Task.Result> recomputeUserCurrentQuotas(Context context, Username username) {
+    private Mono<Task.Result> recomputeQuotasOfUser(List<QuotaComponent> quotaComponents, Context context, Username username) {
+        Map<QuotaComponent, Mono<Task.Result>> map = ImmutableMap.of(QuotaComponent.MAILBOX, recomputeMailboxUserCurrentQuotas(context, username),
+            QuotaComponent.JMAP_UPLOADS, recomputeJMAPCurrentUploadUsage(username));
+        if (quotaComponents.isEmpty()) {
+            return Flux.merge(map.values()).reduce(Task.Result.COMPLETED, Task::combine);
+        } else {
+            return Flux.fromIterable(quotaComponents).flatMap(map::get).reduce(Task.Result.COMPLETED, Task::combine);
```

Comment:
What if we ask for a component not in the map?

At the very least, this line of code is tightly coupled with input validation.

Maybe something like 

```suggestion
            return Flux.fromIterable(quotaComponents)
                .flatMap(c -> mapgetOrElse(c, Mono.just(PARTIAL))
                .reduce(Task.Result.COMPLETED, Task::combine);
```